src/pmdas/amdgpu/amdgpu.c

```diff
@@ -0,0 +1,547 @@
+    return sts;
+  }
+
+  device_count = pcp_amdgpuinfo.numcards;
+  pmNotifyErr(LOG_WARNING, "setup_gcard_indom: got %d cards", device_count);
+
+  pcp_amdgpuinfo.indom = idp;
+  pcp_amdgpuinfo.indom->it_numinst = 0;
+  pcp_amdgpuinfo.indom->it_set =
+      (pmdaInstid *)calloc(device_count, sizeof(pmdaInstid));
+
+  if (!pcp_amdgpuinfo.indom->it_set) {
+    pmNoMem("gcard indom", device_count * sizeof(pmdaInstid), PM_RECOV_ERR);
+    free(pcp_amdgpuinfo.devs);
+    return -ENOMEM;
+  }
+
+  if ((pcp_amdgpuinfo.info = (amdgpu_info_t *)calloc(
+           device_count, sizeof(amdgpu_info_t))) == NULL) {
+    pmNoMem("gcard values", device_count * sizeof(amdgpu_info_t), PM_RECOV_ERR);
+    free(pcp_amdgpuinfo.devs);
+    free(pcp_amdgpuinfo.indom->it_set);
+    return -ENOMEM;
+  }
+
+  for (i = 0; i < device_count; i++) {
+    drmDevicePtr dev = pcp_amdgpuinfo.devs[i];
+
+    pcp_amdgpuinfo.indom->it_set[i].i_inst = i;
+    pmsprintf(gpuname, sizeof(gpuname), "gpu%d", i);
+    if ((name = strdup(gpuname)) == NULL) {
+      pmNoMem("gcard instname", strlen(gpuname), PM_RECOV_ERR);
+      while (--i)
+        free(pcp_amdgpuinfo.indom->it_set[i].i_name);
+      free(pcp_amdgpuinfo.devs);
+      free(pcp_amdgpuinfo.indom->it_set);
+      free(pcp_amdgpuinfo.info);
+      return -ENOMEM;
+    }
+    pcp_amdgpuinfo.indom->it_set[i].i_name = name;
+
+    /* Get static values */
+    if (localDRMDeviceGetGPUInfo(dev, &pcp_amdgpuinfo.info[i].gpu_info)) {
+	pcp_amdgpuinfo.info[i].failed[AMDGPU_MEMORY_CLOCK_MAX] = 1;
+	pcp_amdgpuinfo.info[i].failed[AMDGPU_GPU_CLOCK_MAX] = 1;
+    }
+  }
+
+  pcp_amdgpuinfo.indom->it_numinst = device_count;
+  return 0;
+}
+
+static int refresh(pcp_amdgpuinfo_t *amdgpuinfo) {
+  drmMemory_t memory = {0};
+  int i, j;
+
+  if (!drm_initialized) {
+      setup_gcard_indom();
+      drm_initialized = 1;
+  }
+
+  for (i = 0; i < amdgpuinfo->numcards && i < amdgpuinfo->maxcards; i++) {
+    amdgpu_info_t *info = &amdgpuinfo->info[i];
+    drmDevicePtr dev = amdgpuinfo->devs[i];
+
+    info->cardid = i;
+    for (j = 0; j < AMDGPU_METRIC_COUNT; j++)
+      info->failed[j] = 0;
+    if (localDRMDeviceGetName(dev, info->name))
+      info->failed[AMDGPU_CARDNAME] = 1;
+    if (localDRMDeviceGetMemoryInfo(dev, &memory)) {
+      info->failed[AMDGPU_MEMUSED] = 1;
+      info->failed[AMDGPU_MEMTOTAL] = 1;
+      info->failed[AMDGPU_MEMFREE] = 1;
+    }
+    if (localDRMDeviceGetMemoryClock(dev, &info->mem_clock)) {
+      info->failed[AMDGPU_MEMORY_CLOCK] = 1;
+    }
+    if (localDRMDeviceGetGPUClock(dev, &info->gpu_clock)) {
+      info->failed[AMDGPU_GPU_CLOCK] = 1;
+    }
+    if (localDRMDeviceGetTemperature(dev, &info->temperature)) {
+      info->failed[AMDGPU_TEMPERATURE] = 1;
+    }
+    if (localDRMDeviceGetGPULoad(dev, &info->load)) {
+      info->failed[AMDGPU_GPU_LOAD] = 1;
+    }
+    if (localDRMDeviceGetGPUAveragePower(dev, &info->avg_pwr)) {
+      info->failed[AMDGPU_GPU_AVG_PWR] = 1;
+    }
```

Comment:
The code pattern used here equates to "for every fetch, fetch all the metrics available".  Sometimes, its not ideal (e.g. when sometimes only one or two metrics are fetched, but maybe some library calls are relatively expensive) - if this is not the desired behaviour (i.e. if you want to use the minimum number of library API calls) you can use the fetch/refresh pattern from src/pmdas/linux/pmda.c with a need_refresh array and carefully setup pmID clusters.
But, up to you - it may not be an issue worth spending time on here.